stdlib/public/Reflection/Sources/_Runtime/ImageInspection.cpp

```diff
@@ -0,0 +1,86 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+#include <stdint.h>
+
+#include "swift/Runtime/Config.h"
+#include "swift/Threading/Once.h"
+
+extern "C"
+void registerAttributes(const void *section, intptr_t size);
+
+//===----------------------------------------------------------------------===//
+// Mach-O Image Inspection
+//===----------------------------------------------------------------------===//
+
+#if defined(__MACH__)
+
+#include <mach-o/dyld.h>
+#include <mach-o/getsect.h>
+
+#if __POINTER_WIDTH__ == 64
+typedef mach_header_64 mach_header_platform;
+#else
+typedef mach_header mach_header_platform;
+#endif
+
+void lookupSection(const struct mach_header *header, const char *segment,
+                   const char *section,
+                    void (*registerFunc)(const void *, intptr_t)) {
+  unsigned long size = 0;
+
+  auto sectionData = getsectiondata(
+    reinterpret_cast<const mach_header_platform *>(header), segment, section,
+    &size);
+
+  registerFunc(sectionData, size);
+}
+
+void imageFunc(const struct mach_header *header, intptr_t size) {
+  lookupSection(header, "__TEXT", "__swift5_rattrs", registerAttributes);
+}
+
+extern "C" SWIFT_CC(swift)
+void snapshot() {
+  static swift::once_t token;
+  swift::once(token, []{
+    _dyld_register_func_for_add_image(imageFunc);
+  });
+}
+
+#else
+
+#include "swift/shims/MetadataSections.h"
+
+extern "C"
+void swift_enumerateAllMetadataSections(
+  bool (* body)(const swift::MetadataSections *sections, void *context),
+  void *context
+);
+
+bool imageFunc(const swift::MetadataSections *sections, void *context) {
+  if (sections->version >= 3) {
+    return false;
+  }
+
+  auto runtimeAttrs = sections->swift5_runtime_attributes;
+  registerAttributes(reinterpret_cast<const void *>(runtimeAttrs.start),
+                     runtimeAttrs.length);
+
+  return false;
+}
+
+extern "C" SWIFT_CC(swift)
+void snapshot() {
```

Comment:
This function is in the global C namespace as "snapshot". Did you want "swift_snapshot" or something like that?

Or perhaps:

```c
swift_enumerateRuntimeAttributesSections(void (* body)(const void *start, size_t length, void *context), void *context);
```

And have it call directly through to `swift_enumerateAllMetadataSections()` on Linux, and walk the current snapshot on Darwin? That way, you don't need to expose `registerAttributes` across language boundaries.